File: scripts/gen_specs.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import importlib
import json
import logging
import pkgutil
import sys
from collections.abc import Sequence
from dataclasses import MISSING, Field, asdict, fields
from pathlib import Path
from typing import Any

import mirage.commands.builtin
from mirage.commands.builtin.generic_bind.adapter import CommandIO
from mirage.commands.config import RegisteredCommand
from mirage.commands.spec import SPECS
from mirage.commands.spec.types import CommandSpec, Operand, Option
from mirage.resource.base import BaseResource
from mirage.resource.registry import REGISTRY, resolve_class

logger = logging.getLogger(__name__)

OUT = Path(__file__).resolve().parent.parent / "spec" / "python" / "general"

BUILTIN = Path(mirage.commands.builtin.__file__).resolve(
).parent  # type: ignore[arg-type]

# Slots holding a configuration value rather than an operation. Everything
# else on the adapter is a wired operation, reported by name.
IO_VALUE_FIELDS = frozenset({"local", "max_glob_matches", "max_du_entries"})


def _walk_pkg(pkg: Any) -> list[str]:
    """Import every builtin command module, reporting the ones that failed.

    A module that will not import registers nothing, so its resources
    silently vanish from the dump. That reads as a legitimate deletion in
    the committed spec rather than as the under-provisioned environment it
    actually is, so the caller turns any failure into a hard error.

    Args:
        pkg (Any): the package to walk.
    """
    failed: list[str] = []
    for _finder, name, _ispkg in pkgutil.walk_packages(pkg.__path__,
                                                       pkg.__name__ + "."):
        try:
            importlib.import_module(name)
        except ImportError as e:
            logger.debug("skip %s: %s", name, e)
            failed.append(f"{name}: {e}")
    return failed


def _collect_registrations() -> dict[str, list[RegisteredCommand]]:
    out: dict[str, list[RegisteredCommand]] = {}
    seen: set[int] = set()
    for mod_name, mod in list(sys.modules.items()):
        if mod is None or not mod_name.startswith("mirage.commands."):
            continue
        candidates = list(vars(mod).values())
        commands = getattr(mod, "COMMANDS", None)
        if isinstance(commands, Sequence):
            candidates.extend(commands)
        for attr in candidates:
            if id(attr) in seen:
                continue
            seen.add(id(attr))
            rcs = ([attr] if isinstance(attr, RegisteredCommand) else getattr(
                attr, "_registered_commands", None))
            if not rcs:
                continue
            for rc in rcs:
                out.setdefault(rc.name, []).append(rc)
    return out


def _by_resource(rcs: list[RegisteredCommand]) -> dict[str, Any]:
    """Per-registration metadata, keyed by resource.

    The union flags below cannot say *which* resource carries a provision,
    an aggregate, the write flag or a filetype, so dropping one backend's
    provision while another keeps it leaves every union unchanged. Key the
    same facts by resource so the parity check sees that difference.

    Args:
        rcs (list[RegisteredCommand]): every registration for one command.
    """
    out: dict[str, Any] = {}
    for rc in rcs:
        entry = out.setdefault(
            rc.resource if rc.resource is not None else "", {
                "has_provision": False,
                "has_aggregate": False,
                "has_write": False,
                "filetypes": set(),
            })
        entry["has_provision"] |= rc.provision_fn is not None
        entry["has_aggregate"] |= rc.aggregate is not None
        entry["has_write"] |= bool(rc.write)
        if rc.filetype is not None:
            entry["filetypes"].add(rc.filetype)
    return out


def _meta_for(rcs: list[RegisteredCommand]) -> dict[str, Any]:
    resources = sorted({rc.resource for rc in rcs if rc.resource is not None})
    filetypes = sorted({rc.filetype for rc in rcs if rc.filetype is not None})
    return {
        "has_provision": any(rc.provision_fn is not None for rc in rcs),
        "has_aggregate": any(rc.aggregate is not None for rc in rcs),
        "has_write": any(rc.write for rc in rcs),
        "resources": resources,
        "filetypes": filetypes,
        "by_resource": _by_resource(rcs),
    }


def _default(o: object) -> object:
    if isinstance(o, (set, frozenset)):
        return sorted(o)
    raise TypeError(f"unserializable: {type(o)}")


def _default_of(f: Field) -> Any:
    if f.default_factory is not MISSING:
        return f.default_factory()
    return f.default


def _prune(payload: dict[str, Any], cls: type) -> dict[str, Any]:
    """``payload`` without the fields ``cls`` would have defaulted anyway.

    A spec dump is a cross-language contract, and restating every
    default in all 93 files buries the handful of facts each command
    actually declares. The defaults come from the dataclass rather than
    a second table, so a field added to a spec type cannot fall out of
    step with this. ``type`` survives even at its default, because what
    a token *is* is the first thing a reader looks for.

    The typescript side prunes against a default-constructed instance
    for the same reason; the two must drop exactly the same keys or the
    parity gate reports every command.

    Args:
        payload (dict[str, Any]): one ``asdict`` level, every key
            present.
        cls (type): the dataclass the payload came from.
    """
    kept: dict[str, Any] = {}
    for f in fields(cls):
        value = payload[f.name]
        if f.name != "type" and value == _default_of(f):
            continue
        kept[f.name] = value
    return kept


def _spec_payload(spec: Any) -> dict[str, Any]:
    payload = _prune(asdict(spec), CommandSpec)
    if "options" in payload:
        payload["options"] = [_prune(o, Option) for o in payload["options"]]
    if "positional" in payload:
        payload["positional"] = [
            _prune(p, Operand) for p in payload["positional"]
        ]
    if payload.get("rest") is not None:
        payload["rest"] = _prune(payload["rest"], Operand)
    return payload


def _emit_one(name: str, spec: Any, rcs: list[RegisteredCommand]) -> None:
    payload = _spec_payload(spec)
    payload["_meta"] = _meta_for(rcs)
    path = OUT / f"{name}.json"
    path.write_text(
        json.dumps(payload, indent=2, sort_keys=True, default=_default) + "\n")


def _capabilities() -> dict[str, dict[str, Any]]:
    """Per-resource behavior values, read off the class, never an instance.

    Registry membership only says a backend can be built. How it behaves
    once mounted is a second hand-maintained surface that drifted just as
    quietly: python kept the 600 s ``index_ttl`` default for postgres and
    mongodb where typescript pins 0, so an ``ls`` of a live schema could
    be ten minutes stale. ``storage_id`` and ``statfs`` are reported as
    "does this class override the base" rather than by value, because the
    base answers are per-instance identity and UNKNOWN.
    """
    out: dict[str, dict[str, Any]] = {}
    for name in sorted(REGISTRY):
        cls = resolve_class(REGISTRY[name].resource_path)
        out[name] = {
            "index_ttl": cls.index_ttl,
            "caches_reads": cls.caches_reads,
            "supports_snapshot": cls.SUPPORTS_SNAPSHOT,
            "sizes_always_known": cls.SIZES_ALWAYS_KNOWN,
            "storage_id": cls.storage_id is not BaseResource.storage_id,
            "statfs": cls.statfs is not BaseResource.statfs,
        }
    return out


def _command_io() -> dict[str, dict[str, Any]]:
    """The wired ``CommandIO`` slots per backend command package.

    The adapter's slot set is a hand-filled literal that no gate reads, so
    a backend can omit ``du`` or ``find`` and quietly fall back to the
    capped readdir walk while its twin pushes the work down to the API.
    Dumping the key set turns that omission into a spec diff.
    """
    out: dict[str, dict[str, Any]] = {}
    for path in sorted(BUILTIN.glob("*/io.py")):
        backend = path.parent.name
        mod = importlib.import_module(f"mirage.commands.builtin.{backend}.io")
        io = getattr(mod, "IO", None)
        if not isinstance(io, CommandIO):
            continue
        slots = sorted(f.name for f in fields(CommandIO)
                       if f.name not in IO_VALUE_FIELDS
                       and getattr(io, f.name) is not None)
        out[backend] = {
            "slots": slots,
            "local": io.local,
            "max_glob_matches": io.max_glob_matches,
            "max_du_entries": io.max_du_entries,
        }
    return out


def _configs() -> dict[str, dict[str, Any] | None]:
    """Per-resource config field sets, read off the pydantic model.

    Registry membership says a backend can be built; ``capabilities`` says
    how it behaves; this says what it can be *told*. A YAML block is
    written once in python's snake_case and has to reach both
    implementations, and nothing compared the two field sets: TypeScript
    accepted a Google ``access_token`` it never read, dropped an ``oci``
    ``path_style`` its rename map had lost, and refused an ``ntn``
    ``api_version`` python had declared. Keys are wire names -- the alias
    when a field carries one (``DatabricksVolumeConfig.schema_name`` is
    ``schema`` on the wire) -- so the parity check compares what a config
    block says, not what an attribute is called. A resource whose entry
    names no config class (ram, disk, redis take raw kwargs) dumps null.

    Returns:
        dict[str, dict[str, Any] | None]: per registry name, ``fields``
            mapping each wire name to whether it is required.
    """
    out: dict[str, dict[str, Any] | None] = {}
    for name in sorted(REGISTRY):
        entry = REGISTRY[name]
        ref = entry.config_path
        if ref is None:
            ref = getattr(resolve_class(entry.resource_path), "CONFIG_CLS",
                          None)
        if ref is None:
            out[name] = None
            continue
        cls = resolve_class(ref) if isinstance(ref, str) else ref
        out[name] = {
            "fields": {
                (field.alias or fname): {
                    "required": field.is_required()
                }
                for fname, field in cls.model_fields.items()
            }
        }
    return out


def _emit_resources(registry: dict[str, list[RegisteredCommand]]) -> None:
    """Dump the two resource-name sets the parity gate compares.

    ``registry`` is what ``build_resource`` can construct by name — the
    hand-maintained table workspace YAML and snapshots go through.
    ``command_resources`` is what the spec tree already knew: every
    resource registering at least one builtin command. A name in the
    second but not the first registers commands yet cannot be mounted by
    name, which is how SharePoint stayed unconstructible in python while
    appearing in every command's ``_meta``.

    Args:
        registry (dict[str, list[RegisteredCommand]]): registrations keyed
            by command name, as collected for the spec dump.
    """
    command_resources: set[str] = set()
    for rcs in registry.values():
        for rc in rcs:
            if rc.resource is not None:
                command_resources.add(str(rc.resource))
    payload = {
        "registry": sorted(REGISTRY),
        "command_resources": sorted(command_resources),
        "capabilities": _capabilities(),
        "command_io": _command_io(),
        "configs": _configs(),
    }
    path = OUT.parent / "resources.json"
    path.write_text(json.dumps(payload, indent=2, sort_keys=True) + "\n")
    print(f"emitted {len(payload['registry'])} registry names to {path}")


def main() -> None:
    failed = _walk_pkg(mirage.commands.builtin)
    if failed:
        raise SystemExit(
            "command modules failed to import, so their registrations would "
            "be missing from the dump:\n  " + "\n  ".join(failed) +
            "\n\nInstall the optional dependencies first:\n"
            "  cd python && uv sync --all-extras --no-extra camel")
    registry = _collect_registrations()
    OUT.mkdir(parents=True, exist_ok=True)
    for name, spec in sorted(SPECS.items()):
        _emit_one(name, spec, registry.get(name, []))
    print(f"emitted {len(SPECS)} specs to {OUT}")
    _emit_resources(registry)


if __name__ == "__main__":
    main()

File: typescript/scripts/gen-specs.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { mkdirSync, readFileSync, readdirSync, writeFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { fileURLToPath } from 'node:url'

import * as Browser from '@struktoai/mirage-browser'
import * as Node from '@struktoai/mirage-node'

import {
  CommandSpec as SpecClass,
  Operand as OperandClass,
  Option as OptionClass,
  SPECS,
} from '@struktoai/mirage-core/commands/spec/index'
import { DEFAULT_MAX_DU_ENTRIES } from '@struktoai/mirage-core/commands/builtin/generic/du'
import { DEFAULT_MAX_GLOB_MATCHES } from '@struktoai/mirage-core/utils/glob_walk'

import type { CommandSpec, Operand, Option } from '@struktoai/mirage-core/commands/spec/index'
import type { RegisteredCommand } from '@struktoai/mirage-core/commands/config'

import {
  type Capabilities,
  type CommandIoFacts,
  type ConfigFacts,
  capabilitiesOf,
  collectClasses,
  commandIoFacts,
  configFacts,
  registryClasses,
} from './resource_facts.ts'

const __dirname = resolve(fileURLToPath(import.meta.url), '..')
const SPEC_ROOT = resolve(__dirname, '..', '..', 'spec', 'typescript')
const PACKAGES = resolve(__dirname, '..', 'packages')

// Bespoke Google Workspace API passthroughs. They register command names that
// are not in SPECS, so they contribute nothing to the spec dump and stay
// internal to the gws resource rather than being re-exported.
const UNEXPORTED_COMMAND_GROUPS: ReadonlySet<string> = new Set([
  'GWS_DOCS_API_COMMANDS',
  'GWS_DRIVE_API_COMMANDS',
  'GWS_GMAIL_API_COMMANDS',
  'GWS_SHEETS_API_COMMANDS',
  'GWS_SLIDES_API_COMMANDS',
])

type ModuleBag = Record<string, unknown>

function commandGroupDirs(pkg: string): { dir: string; groups: string[] }[] {
  const root = resolve(PACKAGES, pkg, 'src', 'commands', 'builtin')
  const out: { dir: string; groups: string[] }[] = []
  for (const entry of readdirSync(root, { withFileTypes: true })) {
    if (!entry.isDirectory()) continue
    let source: string
    try {
      source = readFileSync(resolve(root, entry.name, 'index.ts'), 'utf8')
    } catch (err) {
      // A directory with no index.ts declares no command group. Any other
      // read failure means the scan is incomplete, which is exactly when
      // the reachability assertion below must not be trusted.
      if ((err as NodeJS.ErrnoException).code === 'ENOENT') continue
      throw err
    }
    const groups = [...source.matchAll(/^export const ([A-Z0-9_]+_COMMANDS)\b/gm)].map(
      (m) => m[1] as string,
    )
    if (groups.length > 0) out.push({ dir: entry.name, groups })
  }
  return out
}

function declaredCommandGroups(pkg: string): string[] {
  return commandGroupDirs(pkg).flatMap((d) => d.groups)
}

// core ships as a module tree, so its groups are read from the modules that
// declare them rather than from whatever the package index happens to name.
// Nothing can go missing here: the directory scan is the source of truth.
async function coreCommandGroups(): Promise<ModuleBag> {
  const bag: ModuleBag = {}
  for (const { dir } of commandGroupDirs('core')) {
    const mod = (await import(`@struktoai/mirage-core/commands/builtin/${dir}/index`)) as ModuleBag
    for (const [key, value] of Object.entries(mod)) {
      if (key.endsWith('_COMMANDS')) bag[key] = value
    }
  }
  return bag
}

// node and browser are still bundled behind a single entry, so their registry
// can only see command groups the package index re-exports. A backend that
// defines its commands but forgets the re-export silently drops out of the
// spec dump (and out of the cross-language parity check with it), so fail
// loudly instead of emitting a quietly incomplete spec.
function assertGroupsReachable(pkgs: readonly string[], modules: ModuleBag[]): void {
  const reachable = new Set(modules.flatMap((m) => Object.keys(m)))
  const missing: string[] = []
  for (const pkg of pkgs) {
    for (const name of declaredCommandGroups(pkg)) {
      if (reachable.has(name) || UNEXPORTED_COMMAND_GROUPS.has(name)) continue
      missing.push(`${name} (packages/${pkg})`)
    }
  }
  if (missing.length > 0) {
    throw new Error(
      `command groups are not re-exported from their package index, so their ` +
        `registrations are invisible to the spec dump: ${missing.join(', ')}`,
    )
  }
}

function collectRegistrations(modules: ModuleBag[]): Record<string, RegisteredCommand[]> {
  const out: Record<string, RegisteredCommand[]> = {}
  for (const mod of modules) {
    for (const [key, value] of Object.entries(mod)) {
      if (!key.endsWith('_COMMANDS') || !Array.isArray(value)) continue
      for (const rc of value as RegisteredCommand[]) {
        if (!out[rc.name]) out[rc.name] = []
        out[rc.name].push(rc)
      }
    }
  }
  return out
}

// The union flags below cannot say *which* resource carries a provision, an
// aggregate, the write flag or a filetype, so dropping one backend's
// provision while another keeps it leaves every union unchanged. Key the same
// facts by resource so the parity check sees that difference.
function byResource(rcs: RegisteredCommand[]): Record<string, unknown> {
  const out: Record<
    string,
    { has_provision: boolean; has_aggregate: boolean; has_write: boolean; filetypes: Set<string> }
  > = {}
  for (const rc of rcs) {
    const key = rc.resource ?? ''
    const entry = (out[key] ??= {
      has_provision: false,
      has_aggregate: false,
      has_write: false,
      filetypes: new Set<string>(),
    })
    entry.has_provision ||= rc.provisionFn !== null
    entry.has_aggregate ||= rc.aggregate !== null
    entry.has_write ||= rc.write
    if (rc.filetype !== null) entry.filetypes.add(rc.filetype)
  }
  return Object.fromEntries(
    Object.entries(out).map(([key, entry]) => [
      key,
      { ...entry, filetypes: [...entry.filetypes].sort() },
    ]),
  )
}

function metaFor(rcs: RegisteredCommand[]): Record<string, unknown> {
  const resources = [
    ...new Set(rcs.map((r) => r.resource).filter((r): r is string => r !== null)),
  ].sort()
  const filetypes = [
    ...new Set(rcs.map((r) => r.filetype).filter((f): f is string => f !== null)),
  ].sort()
  return {
    by_resource: byResource(rcs),
    filetypes,
    has_aggregate: rcs.some((r) => r.aggregate !== null),
    has_provision: rcs.some((r) => r.provisionFn !== null),
    has_write: rcs.some((r) => r.write),
    resources,
  }
}

// A spec dump is a cross-language contract, and restating every default
// in all 93 files buries the handful of facts each command actually
// declares. Anything equal to what a default-constructed instance would
// have carried is dropped, so the defaults come from the class rather
// than a table that could drift from it. `type` survives even at its
// default, because what a token *is* is the first thing a reader looks
// for. Python's `_prune` does the same against its dataclass fields; the
// two must drop exactly the same keys or the parity gate reports every
// command.
function prune(
  full: Record<string, unknown>,
  defaults: Record<string, unknown>,
): Record<string, unknown> {
  const kept: Record<string, unknown> = {}
  for (const [key, value] of Object.entries(full)) {
    if (key !== 'type' && JSON.stringify(value) === JSON.stringify(defaults[key])) continue
    kept[key] = value
  }
  return kept
}

function operandFields(op: Operand): Record<string, unknown> {
  return {
    name: op.name,
    provided_by: [...op.providedBy],
    remainder: op.remainder,
    required: op.required,
    text_when: [...op.textWhen],
    type: op.type,
  }
}

function serializeOperand(op: Operand): Record<string, unknown> {
  return prune(operandFields(op), operandFields(new OperandClass({})))
}

function optionFields(o: Option): Record<string, unknown> {
  return {
    choices: o.choices,
    count: o.count,
    default: o.default,
    description: o.description,
    env: o.env,
    long: o.long,
    metavar: o.metavar,
    multiple: o.multiple,
    numeric_shorthand: o.numericShorthand,
    pair: o.pair,
    required: o.required,
    short: o.short,
    short_value: o.shortValue,
    type: o.type,
    value_optional: o.valueOptional,
  }
}

function serializeOption(o: Option): Record<string, unknown> {
  return prune(optionFields(o), optionFields(new OptionClass({})))
}

function specFields(spec: CommandSpec): Record<string, unknown> {
  return {
    description: spec.description,
    epilog: spec.epilog,
    ignore_tokens: [...spec.ignoreTokens].sort(),
    old_option_style: spec.oldOptionStyle,
    operand_base: spec.operandBase,
    options: spec.options.map(serializeOption),
    positional: spec.positional.map(serializeOperand),
    rest: spec.rest === null ? null : serializeOperand(spec.rest),
  }
}

function serializeSpec(spec: CommandSpec, rcs: RegisteredCommand[]): Record<string, unknown> {
  return {
    ...prune(specFields(spec), specFields(new SpecClass({}))),
    _meta: metaFor(rcs),
  }
}

function sortedStringify(value: unknown): string {
  return JSON.stringify(
    value,
    (_k, v) => {
      if (v && typeof v === 'object' && !Array.isArray(v)) {
        return Object.fromEntries(
          Object.entries(v as Record<string, unknown>).sort(([a], [b]) => a.localeCompare(b)),
        )
      }
      return v
    },
    2,
  )
}

// The two resource-name sets the parity gate compares. `registry` is what
// `buildResource` can construct by name — the hand-maintained table that
// workspace YAML and snapshots go through. `command_resources` is what the
// spec tree already knew: every resource registering at least one builtin
// command. A name in the second but not the first registers commands yet
// cannot be mounted by name, which is how chroma/dify/lancedb/qdrant stayed
// unconstructible in typescript while appearing in every command's `_meta`.
//
// `capabilities` and `command_io` carry the values behind those names.
// Registry membership only says a backend can be built; how it behaves is
// a second hand-maintained surface that drifted just as quietly — Python
// served ten-minute-stale listings of a live postgres schema because its
// `index_ttl` kept the 600 s default where typescript pinned 0, and box's
// `du` slot is wired on one side and absent on the other.
function emitResources(
  name: string,
  knownResources: string[],
  registry: Record<string, RegisteredCommand[]>,
  capabilities: Record<string, Capabilities | null>,
  commandIo: Record<string, CommandIoFacts>,
  configs: Record<string, ConfigFacts | null>,
): void {
  const commandResources = new Set<string>()
  for (const rcs of Object.values(registry)) {
    for (const rc of rcs) if (rc.resource !== null) commandResources.add(rc.resource)
  }
  const payload = {
    registry: [...knownResources].sort(),
    command_resources: [...commandResources].sort(),
    capabilities,
    command_io: commandIo,
    configs,
  }
  const path = resolve(SPEC_ROOT, name, 'resources.json')
  writeFileSync(path, sortedStringify(payload) + '\n')
  console.log(`emitted ${payload.registry.length} registry names to ${path}`)
}

// Every registry name's capability values, read from the class the entry
// constructs. A name whose class cannot be resolved is a hard error: a
// missing row would read as "no divergence here" in the parity gate.
function capabilitiesFor(
  pkgs: readonly string[],
  variantPkg: string,
): Record<string, Capabilities | null> {
  const classes = collectClasses(PACKAGES, pkgs)
  const names = registryClasses(resolve(PACKAGES, variantPkg, 'src', 'resource', 'registry.ts'))
  const out: Record<string, Capabilities | null> = {}
  for (const [resource, className] of [...names].sort(([a], [b]) => a.localeCompare(b))) {
    out[resource] = className === null ? null : capabilitiesOf(className, classes)
  }
  return out
}

function emitVariant(
  name: string,
  pkgs: readonly string[],
  modules: ModuleBag[],
  knownResources: string[],
): void {
  // core is in `pkgs` because its source is scanned for capabilities and
  // CommandIO facts, but only the runtime package is asserted reachable:
  // core's groups came from the directory scan, which cannot miss one.
  assertGroupsReachable([pkgs[pkgs.length - 1] as string], modules)
  const registry = collectRegistrations(modules)
  const outDir = resolve(SPEC_ROOT, name, 'general')
  mkdirSync(outDir, { recursive: true })
  const cmdNames = Object.keys(SPECS).sort()
  for (const cmd of cmdNames) {
    const spec = SPECS[cmd]
    const rcs = registry[cmd] ?? []
    const payload = serializeSpec(spec, rcs)
    writeFileSync(resolve(outDir, `${cmd}.json`), sortedStringify(payload) + '\n')
  }
  console.log(`emitted ${cmdNames.length} specs to ${outDir}`)
  emitResources(
    name,
    knownResources,
    registry,
    capabilitiesFor(pkgs, pkgs[pkgs.length - 1] as string),
    commandIoFacts(PACKAGES, pkgs, {
      maxGlobMatches: DEFAULT_MAX_GLOB_MATCHES,
      maxDuEntries: DEFAULT_MAX_DU_ENTRIES,
    }),
    configFacts(
      resolve(PACKAGES, pkgs[pkgs.length - 1] as string, 'src', 'resource', 'registry.ts'),
      PACKAGES,
    ),
  )
}

async function main(): Promise<void> {
  const core = await coreCommandGroups()
  emitVariant('node', ['core', 'node'], [core, Node as unknown as ModuleBag], Node.knownResources())
  emitVariant(
    'browser',
    ['core', 'browser'],
    [core, Browser as unknown as ModuleBag],
    Browser.knownResources(),
  )
}

await main()
